### PR TITLE
Added `fs` provider and big refactor of tests

### PR DESCRIPTION
**Pull request by mbrevoort**
- The behavior of the memory provider changed when `serialize` is set to `false`. It has been changed to work like the Persist service and only accept strings.
- Added a new test harness so that tests for any provider will be tested consistently.
- Added a test webserver with Restify to act as a Persist service mock and that will work well with the new test harness.

This should at least be a minor version bump, would do a major but this isn't technically 1.0 yet.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -34,7 +34,8 @@ Returns a Beep Boop Perist api:
...
 + `options.debug` - defaults to `false` - if `true` then api calls and errors are logged.
 + `options.logger` - defaults to `null` - Should be an object w/ a `debug` and `error` function.
-+ `options.provider` - defaults to `null`, acceptable values are `"memory"` or `"beepboop"` - this provides a way to override provider selection logic.  If this isn't explicitly set, then the `"beepboop"` provider is used when both `token` and `url` are present.  Otherwise the `"memory"` provider is used.
++ `options.provider` - defaults to `null`, acceptable values are `"memory"`, `"beepboop"`, or `"fs"` - this provides a way to override provider selection logic.  If this isn't explicitly set, then the `"beepboop"` provider is used when both `token` and `url` are present.  Otherwise the `"memory"` provider is used. `"fs"` can be used when running outside of Beep Boop to make data survive restarts. This provider shouldn't be used while running on Beep Boop though since the disks are ephemeral.
++ `options.directory` - defaults to `.persist` under module - Only used with the `fs` filesystem provider (meant testing only) to control where data is stored.
```
Comment (selfcontained) [inline README.md]: "meant for testing only" ?

File: lib/beepboop-provider.js [hunk 2]
```diff
@@ -67,9 +69,14 @@ module.exports = function serviceProvider (config) {
         return cb(new Error('cannot set a value of undefined'))
       }
 
-      var payload = JSON.stringify({ value: wrap(value) })
+      if (!config.serialize && typeof value !== 'string') {
+        return cb(new Error('value may only be string when serialize=false, not ' + typeof value))
```
Comment (selfcontained) [inline lib/beepboop-provider.js]: nit - could be a string template

File: lib/fs-provider.js [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+      keys = keys || []
+      let total = 0
+      let values = []
+      let anyError = null
+
+      let gather = (err, idx, val) => {
+        anyError = anyError || err
+        values[idx] = val
+        total++
+        if (keys.length === total) {
```
Comment (selfcontained) [inline lib/fs-provider.js:39]: goin' vanilla js here - does feel overkill to pull in some async style lib just for this.  👍

**Review by selfcontained — APPROVED** [review]
a few nits and comments, otherwise LGTM

File: lib/fs-provider.js (hunk 3 above)
Comment (mbrevoort) [reply lib/fs-provider.js:39]: Was trying not to add any new dependencies. `fs-extra` made sense but didn't want to pull in the rest of the world

**Pull request merged**